django/core/cache/backends/redis.py

```diff
@@ -0,0 +1,239 @@
+
+    def _parse_url(self, url):
+        parsed_url = urllib.parse.urlparse(url)
+
+        kwargs = {'host': parsed_url.hostname}
+        if parsed_url.port:
+            kwargs['port'] = parsed_url.port
+
+        if parsed_url.username:
+            kwargs['username'] = parsed_url.username
+
+        if parsed_url.password:
+            kwargs['password'] = parsed_url.password
+
+        return kwargs
+
+    def _get_connection_pool(self, write):
+        index = self._get_connection_pool_index(write)
+
+        params = self._parse_url(self._servers[index])
+        params.update(self._client_kwargs)
+
+        if self._pools[index] is None:
+            if self._servers[index].startswith('unix'):
+                self._pools[index] = self._pool_class(
+                    connection_class=redis.UnixDomainSocketConnection,
+                    path=self._servers[index][5:],
+                    **self._client_kwargs
+                )
+            else:
+                self._pools[index] = self._pool_class(**params)
+
+        return self._pools[index]
+
+    def get_client(self, key='', write=False):
+        pool = self._get_connection_pool(write)
+        return self._client(connection_pool=pool)
+
+    def get(self, key, default):
+        client = self.get_client(key)
+        value = client.get(key)
+        # If key does not exists, return None
+        if value is None:
+            return default
+        else:
+            return self._serializer.loads(value)
+
+    def add(self, key, value, timeout):
+        client = self.get_client(key)
+        return bool(client.set(key, self._serializer.dumps(value), ex=timeout, nx=True))
+
+    def set(self, key, value, timeout):
+        client = self.get_client(key, write=True)
+        client.set(key, self._serializer.dumps(value), ex=timeout)
+
+    def touch(self, key, timeout):
+        client = self.get_client(key, write=True)
+        if timeout is None:
+            return client.persist(key)
+        else:
+            return client.expire(key, timeout)
+
+    def delete(self, key):
+        client = self.get_client(key, write=True)
+        return bool(client.delete(key))
+
+    def clear(self):
+        client = self.get_client(None, write=True)
+        return client.flushall()
+
+    def get_many(self, keys):
+        client = self.get_client(None)
+        ret = client.mget(keys)
+        return {
+            k: self._serializer.loads(v) for k, v in zip(keys, ret) if v is not None
+        }
+
+    def set_many(self, data, timeout):
+        client = self.get_client(None)
+        serialized_data = {k: self._serializer.dumps(v) for k, v in data.items()}
+
+        client.mset(serialized_data)
+
+        for k, _ in data.items():
+            # Setting timeout for each as redis-py does not support timeout with mset
+            if timeout is None:
+                client.persist(k)
+            else:
+                client.expire(k, timeout)
```

Comment:
Based on my analysis in https://github.com/django/django/pull/14437#discussion_r658621668 I'm not sure we need to worry about `.persist()` here either. If [`MSET`](https://redis.io/commands/mset) replaces values like [`SET`](https://redis.io/commands/set) does, they'll be persistent by default. In that case we can simplify this to:

```python
        # Set timeout for each key individually as .mset() doesn't support
        # setting the timeout for all keys at the same time.
        if timeout is not None:
            for key in data:
                client.expire(key, timeout)
```

Obviously `.touch()` will still need to use `PERSIST` if `timeout` is `None`.

Reply:
Yes, by default it sets the key as persistant. Changing this